multilevel irq: output the interrupt level of a node in `gen_defines.py` and simplify multilevel `DT_IRQN`

ycsin opened the pull request:
While the devicetree has everything about the interrupt info of a node, there's no (easy) way to access that information, this results in an overly complicated and error-prone macros for the multilevel IRQ DT macro.

This PR modifies the `gen_defines.py` script to output the interrupt level of a node, added new DT macros to access those and added test for the new DT macros.

With that, the multilevel IRQ DT macros are greatly simplified, and should be good enough util a complete rewrite of our interrupt architecture finally arrives

ycsin committed:
gen_defines: output the interrupt level of a node

Currently it is tedious to know the level of an interrupt for
a node in C. One would have to go through a very complex and
error prone macros to check if there's a parent interrupt
controller & if the controller has an interrupt number and thus
not a pseudo interrupt controller like the one found in
`rv32m1`. The level of a node is required to encode the
Zephyr's multi-level interrupt number

Since it is easier to do it in the `gen_defines` script, let's
do the heavy lifting there so that we can introduce new DT
macros to get the interrupt level very easily later.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
devicetree.h: add macros to get the interrupt level of a node

Added new DT macros to get the interrupt level of a node.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: devicetree: api: test DT_IRQ_LEVEL and DT_INST_IRQ_LEVEL

Added new test for `DT_IRQ_LEVEL` and `DT_INST_IRQ_LEVEL`.

Introduced a new `vnd.cpu-intc` compatible so that we have a
root level interrupt controller that acts as level 1
aggregator, and modified `test_intc` to be level 2 aggregator,
updated test of `DT_IRQN(TEST_I2C_BUS)` accordingly.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

bjarki-trackunit commented:
The code seems fine, but can I ask what we can use the interrupt level for?

ycsin commented:
Converted to draft as I realized that I can probably spin-off a few more things from the other branch to here

ycsin commented:
> The code seems fine, but can I ask what we can use the interrupt level for?

Currently the encoding of the multilevel interrupt scheme in the `devicetree.h` relied on DT macros to check if a node has parent / grandparent interrupt controller then shift the IRQ number accordingly. With #66707 and this PR, we would be able to achieve the same with much less.

bjarki-trackunit commented:
> > The code seems fine, but can I ask what we can use the interrupt level for?
> 
> Currently the encoding of the multilevel interrupt scheme in the `devicetree.h` relied on DT macros to check if a node has parent / grandparent interrupt controller (and additional check to filter out pseudo interrupt controllers) then shift the IRQ number accordingly. With #66707 and this PR, we would be able to achieve the same with much less.

I see :) Am I correct in assuming we would not need to track levels if we used irq numbers relative to their interrupt controller, instead of a "global" irq number? Like
```
nvic {
        interrupts = <0>; /* nvic irq 0 */
        pic0 {
                interrupts = <0>; /* pic0 irq 0 */
                gpioa {
                        interrupts = <0>; /* gpioa irq 0 */
                };
        };
};
```
and configured them like
`IRQ_CONNECT_INTC(intc, irq, ...)`
instead of
`IRQ_CONNECT(irq, ...)`
?

ycsin commented:
> configured them like
> `INTC_IRQ_CONNECT(intc, irq, ...)`
> instead of
> `IRQ_CONNECT(irq, ...)`
> ?

Not sure about these, aren't we only have the `IRQ_CONNECT`? The controller information is currently encoded into the `irq`

bjarki-trackunit commented:
@ycsin Currently we only have `IRQ_CONNECT` yes. I'm proposing we replace it with a new macro `INTC_IRQ_CONNECT()`, whereby encoding the interrupt controller and its level is not necessary.

From the perspective of an IRQ, you only need to know the direct interrupt controller of said IRQ. It does not matter which level the interrupt controller is relative to whichever core the top level interrupt controller belong to. The interrupt controller itself only needs to know its direct parent, etc. etc.

Observe the example I gave in the previous commit, each interrupt controller only needs to know of its own interrupt controller (except for the top level one). IRQs will propagate up the chain from the sensor to the top level controller in a scalable manner.

Can we currently handle multiple interrupt controllers on the same level? like multiple cores, or a bunch of GPIO controllers?

ycsin commented:
> Can we currently handle multiple interrupt controllers on the same level?

Not sure about the other questions, but the current multilevel scheme is able to handle multiple interrupt controllers at the same level, provided that the driver supports multi-instance, i.e. [`intc_plic`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/interrupt_controller/intc_plic.c). Currently there is no upstream board with more than 1 interrupt controllers, I'm trying to add a virtual one in #65500

> Currently we only have `IRQ_CONNECT` yes. I'm proposing we replace it with a new macro `INTC_IRQ_CONNECT()`, whereby encoding the interrupt controller and its level is not necessary.

That's definitely the correct direction, when that is done this PR might not even be relevant anymore, since a client like `uart` would be registering itself with the parent interrupt controller, there's probably no concept of 'levels'.

bjarki-trackunit commented:
> That's definitely the correct direction, when that is done this PR might not even be relevant anymore, since a client like `uart` would be registering itself with the parent interrupt controller, there's probably no concept of 'levels'.

Really glad to hear that :) I will pursue this direction a bit further

ycsin committed:
devicetree.h: rework multilevel interrupt encoding macros

Now that we have a very easy way to get the interrupt
controller and the interrupt level of a node, we can
simplify the multilevel interrupt encoding macros.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: devicetree: api: align `DT_IRQN` test

Not sure why the `DT_IRQN` tests are tabbed, fix that.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented:
@ycsin - is this one relevant anymore?

ycsin commented:
> @ycsin - is this one relevant anymore?

still relevant for current interrupt scheme

andyross reviewed (dismissed):
No particular complaint here.  This is rework in the direction I'd like to see, where there's a single framework that derives all the build/runtime interrupt configuration from devicetree without the need for boilerplate or other nonsense by SOC and app code.

And this is just adding a feature in a straightforward way, which seems reasonable to me and more mergeable than a rewrite.  So +1 from me, but it's for sure going to collide with the various rework and may itself need to be reworked.

bjarki-andreasen reviewed (dismissed):
So much cleaner than before, well done :)

gmarull reviewed (dismissed):
lgtm

ycsin commented on include/zephyr/devicetree.h:
```diff
@@ -2537,55 +2545,25 @@
+	 IRQ_TO_L2(DT_IRQ(DT_IRQ_INTC(node_id), irq)) |                                            \
+	 DT_IRQ(DT_IRQ_INTC(DT_IRQ_INTC(node_id)), irq))
+/* DT helper macro for the macros above */
+#define DT_IRQN_LVL_INTERNAL(node_id, idx, level) DT_CAT3(DT_IRQN_L, level, _INTERNAL)(node_id, idx)
```
I noticed that `DT_CAT` macros can be used to avoid manual redirection, so forced pushed to change this line from:

```c
#define DT_IRQN_LVL_IMPL_INTERNAL(node_id, idx, level) DT_IRQN_L##level##_INTERNAL(node_id, idx)
#define DT_IRQN_LVL_INTERNAL(...) DT_IRQN_LVL_IMPL_INTERNAL(__VA_ARGS__)
```

May I have your approvals again? (if CI passes) @andyross @bjarki-trackunit @cfriedt @gmarull

gmarull reviewed (commented):
forgot to mention that you should update `macros.bnf` file.

ycsin commented:
> forgot to mention that you should update `macros.bnf` file.

fixed, thanks
